be/src/exec/stream/aggregate/agg_state_data.cpp

```diff
@@ -0,0 +1,312 @@
+                DCHECK_LT(table_idx, result_chunk->num_columns());
+                // id or index?
+                auto column = result_chunk->get_column_by_index(table_idx);
+                DCHECK_EQ(1, column->size());
+                _agg_function->merge(_agg_fn_ctx, column.get(), agg_group_state[i] + _agg_state_offset, 0);
+            } else {
+                auto status = result_chunk_or.status();
+                DCHECK(status.is_end_of_file());
+                if (!status.is_end_of_file()) {
+                    return status;
+                }
+            }
+        }
+    }
+    DCHECK(j == result_chunks.size());
+    return Status::OK();
+}
+
+// Restore the specific group_by_keys + agg_values's detail into the state.
+Status AggStateData::allocate_retract_state(size_t chunk_size, const std::vector<uint8_t>& keys_not_in_map,
+                                            const vectorized::Column* raw_column,
+                                            const std::vector<ChunkPtrOr>& result_chunks,
+                                            const Buffer<AggGroupStatePtr>& agg_group_state) const {
+    int32_t j = 0;
+    for (auto i = 0; i < chunk_size; i++) {
+        DCHECK_LT(j, chunk_size);
+        auto& result_chunk_or = result_chunks[j++];
+        if (result_chunk_or.ok()) {
+            auto result_chunk = result_chunk_or.value();
+            auto& result_columns = result_chunk->columns();
+            DCHECK_EQ(result_columns.size(), 1);
+            std::vector<const vectorized::Column*> columns = {raw_column, result_columns[0].get()};
+#ifdef BE_TEST
+            for (auto& col : columns) {
+                VLOG_ROW << "[allocate_restract_state] col:" << col->debug_string();
+            }
+#endif
+            _agg_function->restore_detail(_agg_fn_ctx, i, columns, agg_group_state[i] + _agg_state_offset);
+        } else {
+            auto status = result_chunk_or.status();
+            DCHECK(status.is_end_of_file());
+            if (!status.is_end_of_file()) {
+                return status;
+            }
+        }
+    }
+    DCHECK(j == result_chunks.size());
+    return Status::OK();
+}
+
+Status AggStateData::process_chunk(size_t chunk_size, const std::vector<ChunkPtrOr>& result_chunks,
+                                   const Buffer<uint8_t>& keys_not_in_map, const StreamRowOp* ops,
+                                   std::vector<std::vector<const vectorized::Column*>>& raw_columns,
+                                   const Buffer<vectorized::AggDataPtr>& agg_group_state) const {
+    auto* columns = (raw_columns[agg_func_id()]).data();
+    for (int i = 0; i < chunk_size; i++) {
+        auto op = ops[i];
+        switch (op) {
+        case StreamRowOp::INSERT:
+        case StreamRowOp::UPDATE_BEFORE:
+            _agg_function->update(_agg_fn_ctx, columns, agg_group_state[i] + _agg_state_offset, i);
+            break;
+        case StreamRowOp::DELETE:
+        case StreamRowOp::UPDATE_AFTER:
```

Comment:
Need to check if `UPDATE_BEFORE` and `UPDATE_AFTER` adjacent?

Reply:
Why  UPDATE_BEFORE and UPDATE_AFTER must be adjacent?